Refactor ResourceAddresses in order to support generic Strings instead of URIs

irina-mitrea-luxoft opened the pull request:
New pull request created due to issue with automatic close of previous pull request.
Reference to closed [PR](https://github.com/kaazing/gateway/pull/394)

irina-mitrea-luxoft commented:
@jitsni @sbadugu  Hello, Jitu, I have some questions regarding the current pull request (pull request not finalized, work in progress):

- Where should we perform conversions from Strings to URIs? (at which layer should this be performed; some class names/or any relevant examples would be very useful)

- In classes extending ResourceAddressFactorySpi (e.g. HttpResourceAddressFactorySpi), setAlternateOption methods iterates through resourceFactory items and instantiates URI locations using these resourceFactories. With the current approach, should URIs be instantiated at this stage? What is your suggestion regarding this? (for now, in the checked in code, the instantiation is commented out and the initial String is used, TODOs have been added in order to fix this)

- Also, any other feedback on the pull request at its current initial stage would be very useful

Thank you, Irina.

irina-mitrea-luxoft commented:
@jitsni @sbadugu  Hello, Jitu, I would have some questions regarding the current pull request (pull request not finalized, work in progress):

- Where should we perform conversions from Strings to URIs? (at which layer should this be performed; some class names/or any relevant examples would be very useful)

- In classes extending ResourceAddressFactorySpi (e.g. HttpResourceAddressFactorySpi), setAlternateOption methods iterates through resourceFactory items and instantiates URI locations using these resourceFactories. With the current approach, should URIs be instantiated at this stage? What is your suggestion regarding this? (for now, in the checked in code, the instantiation is commented out and the initial String is used, TODOs have been added in order to fix this)

- Also, any other feedback on the pull request at its current initial stage would be very useful

Thank you, Irina.

jitsni commented:
* I think String to URI conversion happens in the corresponding ResourceAddresFactorySpi classes. I would think ResourceAddress#getResource() would return URI
* ResourceAddressFactorySpi#setAlternateOption can take String if it helps.

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, I have updated the pull request with URIs passed to ResourceAddress constructors. There are no compilation errors at the current stage of the pull request. I have performed a build (with no tests for now) and it was successful.

Pending points:
- There are situations where URIs are available, but need to be passed to methods which receive Strings instead of URIs with the current implementation. There are many such scenarios in which I use java.net.URI.toString() method in order to address this scenario. I am not too happy with this approach, and I am looking into using a better approach. Any suggestion would be most welcome. Relevant example:
> ResourceAddress.java line 219: return factory.newResourceAddress0(newExternalURI.toString(), newResourceURI.toString(), newOptions);
- Refactoring of the URIUtils class is also pending.
- An important point is to check is that tests are working properly (together with performing some manual testing of the gateway - ensuring that the gateway can be started properly).
- Also, please let me know if there is any feedback on the current pull request status.

Thank you, Irina.

jitsni commented on resource.address/http/src/test/java/org/kaazing/gateway/resource/address/http/HttpsResourceAddressFactorySpiTest.java:
```diff
@@ -127,8 +128,8 @@ public void shouldBeAbleToSetAMiddleLevelTransportOption()
 
-        ResourceAddress address = addressFactory.newResourceAddress(URI.create(String.format("https://localhost:4949/path")), inputOptions);
-        verifyTransport(address,  URI.create("ssl://localhost:4949"));
+        ResourceAddress address = addressFactory.newResourceAddress(String.format("https://localhost:4949/path"), inputOptions);
```
String.format is not required (it was there in the original code !)

jitsni commented on transport/spi/src/test/java/org/kaazing/gateway/transport/LoggingFilterTest.java:
```diff
@@ -38,8 +38,6 @@
-
 import org.kaazing.gateway.resource.address.IdentityResolver;
 import org.kaazing.gateway.resource.address.ResourceAddress;
 import org.kaazing.gateway.resource.address.ResourceAddressFactories;
```
Can you delete ResourceAddressFactories and let the test use ResourceAddressFactory.

jitsni commented:
I am still going through the code in IDE. I think there are places where ResourceAddress is used to create some other ResourceAddress (with different scheme, path etc). There is conversion (URI -> String) happening in that case. I think it is ok.

Some cases, the new code doesn't have to create URI (for e.g ResourceAddress creation from inet address)

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, thank you for your feedback. I am currently working on getting the ITests working, because there are failures to them and I want to clarify tests issues. 

After that I will continue with the items mentioned above. Please let me know of more feedback as you go through the code in the IDE, your feedback is very useful. Thank you, Irina.

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, I have checked in a new set of fixes based on ITest failures. 
Please let me know if the overall direction of the pull request seems to be the right one; also please give me more feedback as you go through the IDE code, your feedback is very useful. Thank you, Irina.

jitsni commented on management/src/main/java/org/kaazing/gateway/management/snmp/transport/ManagementUdpTransport.java:
```diff
@@ -61,7 +61,8 @@ public void sendMessage(final Address address, final byte[] message) throws IOEx
                 URI connectURI = new URI("udp://" + udpAddress.getInetAddress().getHostAddress() + ":" + udpAddress.getPort());
```
connectURI can be String

irina-mitrea-luxoft replied:
Adding a change for this with my latest commit (will be pushed to my remote branch in 5 minutes).

irina-mitrea-luxoft replied:
Change checked in.

jitsni commented on resource.address/spi/src/main/java/org/kaazing/gateway/resource/address/Comparators.java:
```diff
@@ -266,7 +266,8 @@ public int compare(ResourceAddress addr1, ResourceAddress addr2) {
                         newResultOptions.setOption(TRANSPORT, transportAddress);
-                        ResourceAddress newResult = addressFactory.newResourceAddress(resourceAddress.getExternalURI(), newResultOptions);
+                        //TODO: Clarify whether this can be implemented in a better manner
+                        ResourceAddress newResult = addressFactory.newResourceAddress(resourceAddress.getExternalURI().toString(), newResultOptions);
```
I think this is OK as ResourceAddress exposes URI and the factory takes String.

jitsni commented:
I think the direction is ok in this PR. So let us get the build to pass. Then
1. Separate PR for network interface syntax functionality
2. A PR for enterprise.gateway based on these changes

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, thank you for your reply. I have just managed to get a successful build which also runs tests with some local changes to my gateway project (some tests are ignored for now on my local setup and they require additional investigation). 

I have noticed that after my refactoring there are some issues with URI strings provided in ITest configs which have neither path, nor trailing slash. Consequently, after my refactoring I get errors to ITests which have URIs provided in \<accept\>s or \<connect\>s with no trailing slashes, e.g.
>
GatewayConfiguration configuration =
                new GatewayConfigurationBuilder()
                    .service()
                        .accept(URI.create("https://localhost:8110"))
                        .connect(URI.create("https://localhost:8080"))

I have started looking into this.

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, fixed the issues regarding the trailing slash. I have just checked in a set of changes that fixed the build together with tests (there are still a couple of tests to be clarified, these are marked with @Ignore("TODO: Clarify this") annotation in my latest commit).

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, fixed the issues regarding the trailing slash: I am now using the getCanonicalURI(uri, true) in the GatewayContextResolver on the input String URI.

Also, I have just checked in a set of changes that fixed the build together with tests (there are still a couple of tests to be clarified, these are marked with @ Ignore("TODO: Clarify this") annotation in my latest commit).

jitsni commented on server/src/main/java/org/kaazing/gateway/server/context/resolve/GatewayContextResolver.java:
```diff
@@ -721,15 +722,17 @@ private void parseProperties(Node parent, ServiceProperties properties) {
         for (String uri : acceptURIs) {
-            URI resolvedURI = resolveURI(getCanonicalURI(uri, true));
+            // TODO: Add initial version of this to TcpResourceAddressSpi
+            String resolvedURI = /*resolveURI(*/(getCanonicalURI(uri, true)).toString();
```
It seems like resolveURI adds default port. For e.g http://localhost/path to http://localhost:80/path
But I am not sure if your change would cause an issue. Also, we can add resolveURI(String)

It seems to me that some of this functionality can go into corresponding ResourceAddressFactorySpi (they can add path and default port). But we will keep the changes small and not go there for now.

irina-mitrea-luxoft replied:
@jitsni Just readded resolveURI in GatewayContextResolver with my latest commit, thank you for your feedback. Initially I planned to add that resolution part to the ResourceAddressFactorySpi, as mentioned in the TODO:
// TODO: Add initial version of this to TcpResourceAddressSpi

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, I have addressed some TODOs which I had entered in the code and done a round of reviewing. There are still three tets which fail:
- HttpDirectoryServiceIT
- TcpAcceptorIT
- TcpConnectorIT
Failing tests are ignored for now with "TODO: Clarify this"
Any feedback on these and on the overall pull request would be very useful. Thank you, Irina.

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, I have removed
>
@ Ignore("TODO:Clarify this")

from the all the above-listed test cases (giving details on them also below:
- TcpIdleTimeout->serverDoesNotCloseWithData
- TcpAcceptorIT->concurrentConnections
- TcpConnectorIT->shouldEstablishConcurrentConnections
- HttpDirectoryServiceIT->testHexEncodedUri,testNoServerHeader,testGetIndexCheckWholeResponse

) since these specific tests seem to fail for me locally without my PR changes (tested that they still fail for me on the develop branch).

Consequently **no tests should be ignored by my pull request now**.
I am currently awaiting your feedback on the PR, thank you.

jitsni commented:
* This branch builds on my machine fine.
* I think the readability can be improved by using local variables in some places. For e.g. 
        return factory.newResourceAddress0(URIUtils.uriToString(newExternalURI), URIUtils.uriToString(newResourceURI), newOptions);
* URIUtils.uriToString() may not be necessary. We won't have any impl other than uri.toString(), right ?
* In places where we don't expect string contain all those chars, we can just delegate to URI.create(). For e.g. SslResourceAddressFactorySpi

irina-mitrea-luxoft commented:
Hello, Jitu
>
This branch builds on my machine fine.

Thank you for your feedback.

>
I think the readability can be improved by using local variables in some places. For e.g. return factory.newResourceAddress0(URIUtils.uriToString(newExternalURI), URIUtils.uriToString(newResourceURI), newOptions);

I will try to identify such places.

>
URIUtils.uriToString() may not be necessary. We won't have any impl other than uri.toString(), right ?

I had decided to add a wrapper over uri.toString() due to the fact that URI has both uri.toString() and uri.toASCIIString() methods and wanted to have a method over the new uri.toString() methods I had added in order to be able to modify this in just one place if needed in the future. 
Should I switch back to simply using uri.toString() and remove the URIUtils.uriToString() method?

>
In places where we don't expect string contain all those chars, we can just delegate to URI.create(). For e.g. SslResourceAddressFactorySpi

Can you please provide some more information on this item?

Thank you,
Irina.

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu
>
This branch builds on my machine fine.

Thank you for your feedback.

>
I think the readability can be improved by using local variables in some places. For e.g. return factory.newResourceAddress0(URIUtils.uriToString(newExternalURI), URIUtils.uriToString(newResourceURI), newOptions);

I will try to identify such places.

>
URIUtils.uriToString() may not be necessary. We won't have any impl other than uri.toString(), right ?

I had decided to add a wrapper over uri.toString() due to the fact that URI has both uri.toString() and uri.toASCIIString() methods and wanted to have a method over the new uri.toString() methods I had added in order to be able to modify this in just one place if needed in the future. 
Should I switch back to simply using uri.toString() and remove the URIUtils.uriToString() method?

>
In places where we don't expect string contain all those chars, we can just delegate to URI.create(). For e.g. SslResourceAddressFactorySpi

Can you please provide some more information on this item? Is your comment in regard to the fact that we are performing the validations:
```
 if (host == null) {
            throw new IllegalArgumentException(format("Missing host in URI: %s", location));
        }
        
if (port == -1) {
            throw new IllegalArgumentException(format("Missing port in URI: %s", location));
        }
        
if (path != null && !path.isEmpty()) {
            throw new IllegalArgumentException(format("Unexpected path \"%s\" in URI: %s", path, location));
        }
```
?
They are present due to the fact that they were in the initial code. Any clarifications regarding this would be useful.

Thank you,
Irina.

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu. I have added a refactoring to the PR for situations where extracting URI.create as local variables would improve readability.
I have also added static imports for URIUtils in several classes.
Please provide your feedback on this and on the [comment above](https://github.com/kaazing/gateway/pull/395#issuecomment-174855153).

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu. I have added a refactoring to the PR for situations where extracting URI.create() as local variables would improve readability.
I have also added static imports for URIUtils methods in several classes.

Please provide your feedback on this and on the [comment above](https://github.com/kaazing/gateway/pull/395#issuecomment-174855153). Thank you.

jitsni commented:
I was saying that the following may be better. We don't expect location and original strings to have any network interface related syntax in SslResourceAddressFactorySpi. We can directly use URI. I think this applies to many places.
```
    protected SslResourceAddress newResourceAddress0(String original, String location) {
        URI locationUri = URI.create(original);
        String host = locationUri.getHost(location);
``

irina-mitrea-luxoft commented:
@jitsni Thank you for your clarification, Jitu, as per the discussion on Slack we will rely for path/port/host nullchecks on the URI object validation and no longer throw specific errors from the ResourceAddressFactorySpi classes.

irina-mitrea-luxoft commented:
@jitsni Thank you for your clarification, Jitu, as per the discussion on Slack we will rely for path/port/host nullchecks on the URI object validation and no longer throw specific errors from the ResourceAddressFactorySpi classes.

irina-mitrea-luxoft commented:
@jitsni Thank you for your clarification, Jitu, as per the discussion on Slack we will rely for path/port/host nullchecks on the URI object validation and no longer throw specific errors from the ResourceAddressFactorySpi classes.

Also, considering we will no longer perform validation, the line 
```
String host = locationUri.getHost(location);
```
will also no longer be required and will refactor the existing method to basic ResoourceAddress creation
```
URI uriOriginal = URI.create(original);
URI uriLocation = URI.create(location);
return new SslResourceAddress(this, uriOriginal, uriLocation);
```

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, I have left the validation within the ResoureAddresFactorySpis due to the fact that there are differences within the implementation for different ResourceAddressFactorySpis, e.g.:
-  TcpResourceAddress throws an error
>
 if  (path != null && !path.isEmpty())

- SseResourceAddressFactorySpi throws an error
>
if (path == null || path.length() == 0))

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, I have just merged the latest changes from gateway develop into this branch due to the fact that conflicts had arised due to the latest gateway merges. I have fixed the conflicts and checked in the changes.

Conflicts are likely to occur between the latest gateway merges and this pull request since this pull request impacts a large number of files.

irina-mitrea-luxoft commented:
@jitsni Hello, Jitu, I have also written on Slack, there are situations where there  are some generics maps returned, e.g.
>
Map<MemberId, Map<String, List<String>>> memberIdBalancerMap = factory
               .getMap(HttpBalancerService.MEMBERID_BALANCER_MAP_NAME);

(GatewayManagementBeanImpl, line 392)
In such situations any return type can be specified and these should all be changed accordingly in the source code. Such errors can not be caught at compile time/build, but only at runtime.

We should be very careful with such situations, please check my latest commit as it catches such an issue.

jitsni commented:
IMO, your signature should be fine otherwise the cluster tests would fail. But can we not keep Map<MemberId, Map<URI, List<URI>>> ? 
Post it to engineering channel, someone / @mjolie  should be able to answer it.

mjolie commented:
Hi, not sure what the question is can you clarify ??  We have to be very careful with the cluster and balancer service maps as some of them are created by hazel cast code and we do not have control over that.  We have had 2 very serious customer escalations in this code recently.

irina-mitrea-luxoft commented:
Slack reply from @jitsni [9:41 PM] (#roadmap channel)
This is in 5.0. We are changing some signatures in ResourceAddressFactory. They will take String instead of URI while creating the ResourceAddress objects (this is required since URI cannot deal with network syntax like \@eth0 etc). So gateway server is passing accept, connect, balancer Strings directly to ResourceAddressFactorySpi. Since gateway server is also creating these hazelcast maps, we are wondering to update the signatures to Strings

irina-mitrea-luxoft commented:
Slack reply from @mjolie [11:10 PM] (#Roadmap channel)
@jitsni:  @irina.mitrea  Hi ,  I do not think we will see any compile time issues but I worry about possible runtime issues, we will need to do a lot of testing of the balancer service . Also be sure and test with trace on "ha" trace to be exact there are functions that use the maps that only get called with logging enabled.(edited)

irina-mitrea-luxoft commented:
Reply from @irina-mitrea-luxoft [8:46 AM] (#Roadmap channe;)
@mjolie: @jitsni Hello, Marie, thank you for your feedback and thank you, Jitu, for explaining the issue. There are no compile issues. Runtime problems are the main concern.

Some information on the coverage on the balancer service would be useful, I see that we have some tests under org.kaazing.gateway.service.http.balancer, but I am not sure they cover all the needed scenarios.

Also, I have modified such a map yesterday within GatewayManagementBeanImpl, i.e: 
Map<MemberId, Map<String, List<String>>> memberIdBalancerMap = factory.getMap(HttpBalancerService.MEMBERID_BALANCER_MAP_NAME);

; looking at the existing tests for gateway.management this seems to not be covered by gateway.management tests under the test JMX package.

[Will copy the thread of messages on this topic also within the pull request: https://github.com/kaazing/gateway/pull/395]

mjolie commented:
I have a PR with some new tests that you can run locally , https://github.com/kaazing/gateway/pull/407

mjolie commented:
@irina-mitrea-luxoft @jitsni @sbadugu I have a PR with some new tests that you can run locally , #407, They work locally for me, but are ignored for Travis build until a Robot bug is fixed. I am most concerned about when cluster members go up and down, that the balancer maps have the correct state of each server. You could also ask @daniel-smith-kaazing about the testing environment he used to verify the recent fixes we made to the balancer service for escalations. And remember it is very important that ha trace be enabled and you check the log for class cast exceptions and null pointers. thanks

daniel-smith-kaazing commented:
We did nothing fancy, really. Set up 3 gateways with a balancer service on
all and JMS service on all.
DNS round robin in front handing out IPs in turn for each of the gateways.
Balancer name, DMZ1, DMZ2, DMZ3.
We had scripts that brought each of the nodes up and down and observed in
the ha TRACE if the nodes updated properly when a node joined or left the
cluster.
One variation is to try to make sure at least one node is always up.
That's it.


On Fri, Jan 29, 2016 at 10:40 AM, Marie-Denise Jolie <
notifications@github.com> wrote:

> @irina-mitrea-luxoft <https://github.com/irina-mitrea-luxoft> @jitsni
> <https://github.com/jitsni> @sbadugu <https://github.com/sbadugu> I have
> a PR with some new tests that you can run locally , #407
> <https://github.com/kaazing/gateway/pull/407>, They work locally for me,
> but are ignored for Travis build until a Robot bug is fixed. I am most
> concerned about when cluster members go up and down, that the balancer maps
> have the correct state of each server. You could also ask
> @daniel-smith-kaazing <https://github.com/daniel-smith-kaazing> about the
> testing environment he used to verify the recent fixes we made to the
> balancer service for escalat ions. And remember it is very important that
> ha trace be enabled and you check the log for class cast exceptions and
> null pointers. thanks
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaazing/gateway/pull/395#issuecomment-176902369>.
>



-- 

*Dan Smith*

Senior Support Engineer  |  Phone: 650.960.8155 | Mobile: 408.835.1167
daniel.smith@kaazing.com <dorris.obrien@kaazing.com>

*------------------------------------------------------------*

KAAZING >|<
When Real-time Matters
Kaazing.com <http://www.kaazing.com/>  |  Download
<http://kaazing.com/download.html>  |  Blog <http://blog.kaazing.com/>  |
Twitter <https://twitter.com/kaazing>
*http://kaazing.com <http://kaazing.com/about-us/careers/>*

daniel-smith-kaazing commented:
I would add that IMHO it is essential that three nodes be involved, not
two. I believe that in the past some odd behavior for joining/leaving a
cluster was missed because it may only manifest when there are > 2 nodes.

On Fri, Jan 29, 2016 at 2:00 PM, Dan Smith <daniel.smith@kaazing.com> wrote:

> We did nothing fancy, really. Set up 3 gateways with a balancer service on
> all and JMS service on all.
> DNS round robin in front handing out IPs in turn for each of the gateways.
> Balancer name, DMZ1, DMZ2, DMZ3.
> We had scripts that brought each of the nodes up and down and observed in
> the ha TRACE if the nodes updated properly when a node joined or left the
> cluster.
> One variation is to try to make sure at least one node is always up.
> That's it.
>
>
> On Fri, Jan 29, 2016 at 10:40 AM, Marie-Denise Jolie <
> notifications@github.com> wrote:
>
>> @irina-mitrea-luxoft <https://github.com/irina-mitrea-luxoft> @jitsni
>> <https://github.com/jitsni> @sbadugu <https://github.com/sbadugu> I have
>> a PR with some new tests that you can run locally , #407
>> <https://github.com/kaazing/gateway/pull/407>, They work locally for me,
>> but are ignored for Travis build until a Robot bug is fixed. I am most
>> concerned about when cluster members go up and down, that the balancer maps
>> have the correct state of each server. You could also ask
>> @daniel-smith-kaazing <https://github.com/daniel-smith-kaazing> about
>> the testing environment he used to verify the recent fixes we made to the
>> balancer service for escalat ions. And remember it is very important that
>> ha trace be enabled and you check the log for class cast exceptions and
>> null pointers. thanks
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kaazing/gateway/pull/395#issuecomment-176902369>.
>>
>
>
>
> --
>
> *Dan Smith*
>
> Senior Support Engineer  |  Phone: 650.960.8155 | Mobile: 408.835.1167
> daniel.smith@kaazing.com <dorris.obrien@kaazing.com>
>
> *------------------------------------------------------------*
>
> KAAZING >|<
> When Real-time Matters
> Kaazing.com <http://www.kaazing.com/>  |  Download
> <http://kaazing.com/download.html>  |  Blog <http://blog.kaazing.com/>  |
>  Twitter <https://twitter.com/kaazing>
> *http://kaazing.com <http://kaazing.com/about-us/careers/>*
>



-- 

*Dan Smith*

Senior Support Engineer  |  Phone: 650.960.8155 | Mobile: 408.835.1167
daniel.smith@kaazing.com <dorris.obrien@kaazing.com>

*------------------------------------------------------------*

KAAZING >|<
When Real-time Matters
Kaazing.com <http://www.kaazing.com/>  |  Download
<http://kaazing.com/download.html>  |  Blog <http://blog.kaazing.com/>  |
Twitter <https://twitter.com/kaazing>
*http://kaazing.com <http://kaazing.com/about-us/careers/>*

irina-mitrea-luxoft commented:
Adding reference to QA [story](https://github.com/kaazing/QA-Roadmap/issues/104#issuecomment-181380356) performed by Nicoleta

jitsni commented:
can we close this PR since you another PR (network interface suppor one) builds on this one.

irina-mitrea-luxoft commented:
Adding reference to [PR](https://github.com/kaazing/gateway/pull/437) holding changes for both URI to String refactoring and network interface support.

irina-mitrea-luxoft commented:
As per Jitu's above comment, closing current PR.
The changes present here are now merge-able together with [PR](https://github.com/kaazing/gateway/pull/437) holding changes for both URI to String refactoring and network interface support.
